// use-pouchdb/lib/node/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useView = exports.useFind = exports.useAllDocs = exports.useDoc = exports.usePouch = exports.Provider = void 0;
var context_1 = require("./context");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return context_1.Provider; } });
var usePouch_1 = require("./usePouch");
Object.defineProperty(exports, "usePouch", { enumerable: true, get: function () { return __importDefault(usePouch_1).default; } });
var useDoc_1 = require("./useDoc");
Object.defineProperty(exports, "useDoc", { enumerable: true, get: function () { return __importDefault(useDoc_1).default; } });
var useAllDocs_1 = require("./useAllDocs");
Object.defineProperty(exports, "useAllDocs", { enumerable: true, get: function () { return __importDefault(useAllDocs_1).default; } });
var useFind_1 = require("./useFind");
Object.defineProperty(exports, "useFind", { enumerable: true, get: function () { return __importDefault(useFind_1).default; } });
var useView_1 = require("./useView");
Object.defineProperty(exports, "useView", { enumerable: true, get: function () { return __importDefault(useView_1).default; } });
